src/solvers/refinement/string_refinement.cpp

```diff
@@ -303,70 +323,44 @@ void string_refinementt::concretize_string(const exprt &expr)
 {
   if(refined_string_typet::is_refined_string_type(expr.type()))
   {
-    string_exprt str=to_string_expr(expr);
-    exprt length=get(str.length());
+    const string_exprt str=to_string_expr(expr);
+    const exprt length=get(str.length());
     exprt content=str.content();
     replace_expr(symbol_resolve, content);
     found_length[content]=length;
-    mp_integer found_length;
-    if(!to_integer(length, found_length))
-    {
-      INVARIANT(
-        found_length.is_long(),
-        string_refinement_invariantt("the length of a string should be a "
-          "long"));
-      INVARIANT(
-        found_length>=0,
-        string_refinement_invariantt("the length of a string should be >= 0"));
-      size_t concretize_limit=found_length.to_long();
-      INVARIANT(
-        concretize_limit<=generator.max_string_length,
-        string_refinement_invariantt("string length must be less than the max "
-          "length"));
-      exprt content_expr=str.content();
-      std::vector<exprt> result;
-
-      if(index_set[str.content()].empty())
-        return;
+    const auto string_length=expr_cast<std::size_t>(length);
```

Comment:
We've got two problems: concealing errors because they're returned as error codes and lack of unified way of handling those. Returning error code without any guards is extremely error prone - it's very easy not to check for the error (there are fixes for that, but that should be another task). Also it's not always obvious whether false means success or true. I'm just trying to be consistent with std::dynamic_cast<ref> and boost::lexical_cast - they do throw on failure.

In this case throwing terminates the application, as exception is not handled. That call could be wrapped as invariant but overall result is the same. What I'd like to avoid is having 10 different versions of the "cast exprt to int" each forgetting to check different invariants and having its own set of bugs.

I'd like to swap those true/false returns and exceptions with `boost/std::optional` and/or `expected`, but I'd prefer not to include them here.